contracts/Protocol.sol

```diff
@@ -7,35 +7,42 @@ contract Protocol is IProtocol, Ownable {
         bool exists;
-        uint nextVersion;
+        uint256 nextVersion;
         uint asOfBlock;
-        uint currentVersion;
+        uint256 currentVersion;
     }
 
     mapping (string => DeploymentSubset) deploymentSubsets;
 
-    function deploymentSubsetExists(string memory deploymentSubset) public view returns (bool) {
+    function deploymentSubsetExists(string calldata deploymentSubset) external view returns (bool) {
         return deploymentSubsets[deploymentSubset].exists;
     }
 
+    function createDeploymentSubset(string calldata deploymentSubset, uint256 initialProtocolVersion) external onlyOwner {
+        require(!deploymentSubsets[deploymentSubset].exists, "deployment subset already exists");
+
+        deploymentSubsets[deploymentSubset].currentVersion = initialProtocolVersion;
+        deploymentSubsets[deploymentSubset].nextVersion = initialProtocolVersion;
+        deploymentSubsets[deploymentSubset].asOfBlock = block.number;
+        deploymentSubsets[deploymentSubset].exists = true;
+
+        emit ProtocolVersionChanged(deploymentSubset, initialProtocolVersion, block.number); // TODO different event?
+    }
+
     function setProtocolVersion(string calldata deploymentSubset, uint256 protocolVersion, uint256 asOfBlock) external onlyOwner {
-        if (!deploymentSubsets[deploymentSubset].exists) {
-            require(asOfBlock == 0, "initial protocol version must be from block 0");
-            deploymentSubsets[deploymentSubset].currentVersion = protocolVersion;
-            deploymentSubsets[deploymentSubset].exists = true;
-        } else {
-            uint currentAsOfBlock = deploymentSubsets[deploymentSubset].asOfBlock;
-            if (currentAsOfBlock <= block.number) {
-                deploymentSubsets[deploymentSubset].currentVersion = deploymentSubsets[deploymentSubset].nextVersion;
-            }
-
-            require(asOfBlock > block.number, "protocol update can only be scheduled for a future block");
-            require(protocolVersion > deploymentSubsets[deploymentSubset].currentVersion, "protocol version must be later than current version");
+        require(deploymentSubsets[deploymentSubset].exists, "deployment subset does not exist");
+        require(asOfBlock > block.number, "protocol update can only be scheduled for a future block");
+
+        if (deploymentSubsets[deploymentSubset].asOfBlock <= block.number) {
+            deploymentSubsets[deploymentSubset].currentVersion = deploymentSubsets[deploymentSubset].nextVersion;
         }
+        require(protocolVersion > deploymentSubsets[deploymentSubset].currentVersion, "protocol version must be later than current version");
 
         deploymentSubsets[deploymentSubset].nextVersion = protocolVersion;
         deploymentSubsets[deploymentSubset].asOfBlock = asOfBlock;
 
         emit ProtocolVersionChanged(deploymentSubset, protocolVersion, asOfBlock);
     }
+
+    // TODO what about cancelling a future upgrade?
```

Comment:
this is a good point, maybe we should consider adding a `abortScheduledUpgrade(string calldata deploymentSubset) external onlyOwner`? @OdedWx

Reply:
discussion summary - @OdedWx agrees we should be able to cancel an upgrade by passing the current protocol version  to `setProtocolVersion` - minor detail to resolve - the resulting event should be emitted with the current block number in `asOfBlock` no matter what block number was really sent in. Do we fail this call when sent with `asOfBlock` in the past?

Reply:
It's not a special case - we pass the current protocol version along with a (near-by) future block, the behaviour remains the same. We only need to allow setting to current version.